### ISSUE TITLE
Migration 2.10 -> 2.13.1 drops some suggestions

### ISSUE THREAD

**Issue by stanislav-brabec** [WeblateOrg/weblate#1494]
After migration from 2.10 to 2.13.1, some suggestions changed its text.

### Steps to reproduce
1. Dump database before migration (in my case it was mysql) to weblate210.sql
2. ./manage.py migrate
3. Dump database before migration (in my case it was mysql) to weblate2131.sql
3. Perform commands to compare suggestions:

```shell
grep 'INSERT_INTO `trans_suggestion`' weblate210.sql >before.cut
grep 'INSERT_INTO `trans_suggestion`' weblate2131.sql >after.cut
sed "s/),(/),\\n(/g" <before.cut >before.lines
sed "s/),(/),\\n(/g" <after.cut >after.lines
sed 's/^\(([0-9]*,\)'\''[0-9a-f]*'\'',/\1/' <before.lines >before.comp
sed 's/,[^,()]*)/)/' <after.lines >after.comp
diff -u before.comp after.comp
```

### Actual behaviour
```diff
 (29056,'Atribut LDAP, který odpovídá názvu síťové skupiny.',28,50,154),
 (29058,'Atribut LDAP, který obsahuje jména členů síťové skupiny.',28,50,154),
-(29060,'Atribut LDAP, který obsahuje trojice údajů síťových skupin (hostitel, uživatel, doména).',28,50,154),
+(29060,'Atribut LDAP, který obsahuje trojice údajů síťových skupin (hostitel, uživatel).',28,50,154,4147973960986304414),
 (29062,'Atribut LDAP, který obsahuje identifikátor UUID nebo GUID objektu síťové skupiny protokolu LDAP.',28,50,154),
 (29064,'Třída objektu položky služby v protokolu LDAP.',28,50,154),
 (29066,'Atribut LDAP, který obsahuje název atributů služby a jejich aliasy.',28,50,154),
```

There are many mangled suggestions, but it seems that it has something to do with comma inside brackets.

### Expected behaviour
Not mangle suggestions.

### Server configuration
 * Weblate 2.13.1
 * Python 2.7.12
 * Django 1.10.5
 * six 1.9.0
 * social-auth-core 1.2.0
 * social-auth-app-django 1.1.0
 * django-appconf 1.0.1
 * Translate Toolkit 2.0.0
 * Whoosh 2.7.0
 * defusedxml 0.4.1
 * Git 2.12.0
 * Pillow (PIL) 1.1.7
 * dateutil 2.4.2
 * lxml 3.6.1
 * django-crispy-forms 1.6.1
 * compressor 2.1.1
 * djangorestframework 3.5.3
 * pytz 2015.2
 * pyuca N/A
 * PyYAML 3.11
 * Mercurial 3.8.3
 * git-svn 2.12.0
 * git-review 1.25.0
 * Database backends: django.db.backends.mysql

**Comment by nijel** [comment]
I don't say there can't be problem, but the change for texts with comma inside problem seems to come from your sed expresssion (`sed 's/,[^,()]*)/)/' <after.lines >after.comp`). The table structure has changed, so there are some differences expected (the way units are referenced has been changed from strings to big integer fields). Can you please check what has actually changed in the original SQL dump rather than doing some sed magic, which does further corrupt the data?

**Comment by stanislav-brabec** [comment]
OOPS, my mistake. Invalid sed expression.

After fix to:
```shell
sed 's/,[^,()]*),$/),/' <after.lines >after.comp
```
diffs really disappeared, but I see lost suggestions, exactly like in #1496:
```diff
 (221378,'사용자 스크립트 추가 혹은 변경',274,246,630),
 (221380,'Spice 게스트를 위한 에이전트',274,246,630),
 (221382,'백업 모니터',274,246,630),
-(221428,'Một sự trợ giúp cho cowsay',234,246,NULL),
-(221430,'imageJ cộng đồng java chương trình xử lí hình ảnh',234,246,NULL),
-(221432,'Biên tập (La)TeX',234,246,NULL),
-(221434,'gỡ cài đặt ',234,246,NULL),
-(221436,'15-ô ',234,246,NULL),
-(221442,'hai nghìn không trăm bốn mươi tám \n',234,246,NULL),
-(221446,'Een demo van de Granite toolkit',192,246,634),
 (221448,'Bibliotheek voor het werken met .zip archief',192,244,634),
 (221450,'Minizip is een bibliotheek voor het manipuleren met bestanden van .zip archieven.',192,244,634),
 (221452,'Ontwikkelingsbestanden voor de ZBAR bibliotheek',192,244,634),
```

I just verified, that the string 'Een demo van de Granite toolkit' cannot be found in the whole migrated database dump.

**Comment by stanislav-brabec** [comment]
Here is the pre-migration chunk of the line:
(221446,'27efffb8409b74ab090c07b2bf8a0916','Een demo van de Granite toolkit',192,246,634),
(The whole line is 756594 characters long, so I cannot paste it here.)

**Comment by nijel** [comment]
I don't see any reason why it should remove valid suggestions. So it's likely that these were stale from units no longer present. The cleanup doesn't happen automatically, but only during the [cleanuptrans](https://docs.weblate.org/en/latest/admin/management.html#cleanuptrans) command.

To verify this, you can try finding unit with matching contentsum for one of those suggestions from the original SQL dump. (It's `27efffb8409b74ab090c07b2bf8a0916` in the example you've just posted)

**Comment by stanislav-brabec** [comment]
27efffb8409b74ab090c07b2bf8a0916:
9 occurrences in the 2.10 dump
```sql
INSERT INTO `trans_source` VALUES (...
  (181420,'27efffb8409b74ab090c07b2bf8a0916','2017-01-20 15:11:08',100,374,'',''),...
INSERT INTO `trans_suggestion` VALUES (...
  (221446,'27efffb8409b74ab090c07b2bf8a0916','Een demo van de Granite toolkit',192,246,634),
INSERT INTO `trans_unit` VALUES (...
  (5288564,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19268),
  (5294906,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','أدوات الجرانيت التجريبية',0,1,45,0,0,0,6,100,19270),
  (5301248,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19272),
  (5307590,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19274),
  (5313932,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19276),
  (5320274,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19278),
  (5326616,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','Una demostració de l\'eina Granite',0,1,45,0,0,0,6,100,19280),
  (5332958,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19282),
  5339300,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19284),
  (5345642,'27efffb8409b74ab090c07b2bf8a0916','27efffb8409b74ab090c07b2bf8a0916','/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19286),
```

0 ocurrences in 2.13.1 dump

Note that string 'A demo of the Granite toolkit' exists in 2.13.1 dump
```sql
INSERT INTO `trans_unit` VALUES (...
  (5288564,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19268,-7304718305089163253,-7304718305089163253),
  (5294906,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','أدوات الجرانيت التجريبية',0,1,45,0,0,0,6,100,19270,-7304718305089163253,-7304718305089163253),
  (5301248,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19272,-7304718305089163253,-7304718305089163253),
  (5307590,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19274,-7304718305089163253,-7304718305089163253),
  (5313932,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19276,-7304718305089163253,-7304718305089163253),
  (5320274,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19278,-7304718305089163253,-7304718305089163253),
  (5326616,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','Una demostració de l\'eina Granite',0,1,45,0,0,0,6,100,19280,-7304718305089163253,-7304718305089163253),
  (5332958,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19282,-7304718305089163253,-7304718305089163253),
  (5339300,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19284,-7304718305089163253,-7304718305089163253),
  (5345642,'/usr/share/applications/granite-demo.desktop','Comment(granite-demo.desktop)','','','A demo of the Granite toolkit','','',0,0,45,0,0,0,6,100,19286,-7304718305089163253,-7304718305089163253),
```

**Comment by stanislav-brabec** [comment]
Note that I see inconsistencies even in the original weblate-2.6:
https://l10n.opensuse.org/translate/libzypp/master/zh_TW/?type=suggestions
I clicked to "Strings with suggestions". I see string, but no suggestion.
I will test cleanuptrans before migration next time.

**Comment by nijel** [comment]
But for the Vietnamese (what I guess were those suggestions) the strings are probably not there, that's why the suggestions were removed. The closest one from the po file seems to be here: https://l10n.opensuse.org/translate/desktop-file-translations/update-desktop-files-apps-master/vi/?checksum=2f97a1d84d4d8b8556932f2931344d1e The Vietnamese file seems to miss quite a lot of units in Weblate...

**Comment by stanislav-brabec** [comment]
The unit is there:
https://l10n.opensuse.org/translate/desktop-file-translations/update-desktop-files-apps-master/vi/?sid=7b328dd0-3ca6-11e7-ab39-525400a48651&offset=1390
But I suppose being that Dutch or Danish.
https://l10n.opensuse.org/translate/desktop-file-translations/update-desktop-files-apps-master/nl/?checksum=27efffb8409b74ab090c07b2bf8a0916
(the number is cut-and-paste from the dump above).

But it seems that the string in question was lost by Weblate-2.6 somehow.
The whole update-desktop-files-apps-master contain only a single suggestion, different from this one:
https://l10n.opensuse.org/projects/desktop-file-translations/update-desktop-files-apps-master/
The string cannot be found in po files. Either somebody rejected the suggestion or it become invisible (and dropped on migration) somehow.

**Comment by stanislav-brabec** [comment]
I just did a new dump of the old 2.6 database running on the live system https://l10n.opensuse.org/
Questionable suggestions disappeared from it without any additional action.

Maybe it was an intermediate state, as I did the backup during normal data traffic.

2.13.1 + 1c7cd65 + 74ade3f + 80ed96d did not drop any single suggestion.

If there is a problem, then it is an inconsistency in 2.6. https://github.com/WeblateOrg/weblate/issues/1494#issuecomment-302725172 I will ignore it in 2.6. Let's see whether it disappears on a fully migrated server with cleanuptrans called.
